nixos/tests: add ROCm and amdvlk tests

Flakebi opened the pull request:
###### Description of changes

Testing ROCm and other software like amdvlk is slightly more difficult, because it needs hardware like GPUs to run on.
This PR adds some infrastructure – `make-test-hardware.nix` analogous to `make-test-python.nix` – that makes it easy to write such tests and run them by giving access to the host hardware into the nix build sandbox.

The tests added here are very basic and are meant more for demonstration of how it works (and smoke testing that nothing fundamental is broken).

The added tests can be run with
```bash
$(nix-build -A rocm-opencl-icd.impureTests)
$(nix-build -A amdvlk.impureTests)
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Flakebi committed:
makeImpureTest: init function for hardware tests

Add a function to create tests that use hardware from the host system
like GPUs.

Flakebi committed:
rocm-opencl-icd: add nixos test

Madouura commented:
I've been very sick for a bit now so I can't divert much energy to testing this ATM, but on the surface it looks fine.
I'm concerned on whether we can "legally" include the sandbox changes in the test itself in nixpkgs however.
cc @Mic92 for an authoritative answer on this.

Madouura commented:
Also, ``card0`` is not guaranteed to be an AMD card, I think there should be some logic to ensure it is, and if it is not, if say, ``card1`` exists, test that.

Mic92 commented on nixos/tests/make-test-hardware.nix:
```diff
@@ -0,0 +1,32 @@
+
+    if ${sandboxPathsTest}; then
+      echo 'Run this test as *root* with `--option extra-sandbox-paths '"'${sandboxPathsList}'"'` and a GPU'
+      exit 1
```
I think one issue here is that this test will always show up as "failed" in hydra and ofborg, which is not so nice.

Mic92 replied:
If we are already give up on impurity could we not just have scripts that run outside the nix sandbox and people with the appropriate hardware would run those? Than we also don't need to remember all those sandbox flags.

Mic92 commented on pkgs/development/libraries/rocm-opencl-icd/default.nix:
```diff
@@ -11,6 +11,8 @@ stdenv.mkDerivation rec {
     echo "${rocm-opencl-runtime}/lib/libamdocl64.so" > $out/etc/OpenCL/vendors/amdocl64.icd
   '';
 
+  passthru.tests = { inherit (nixosTests) rocm-opencl; };
```
I could imagine a new attribute called `impureTests`

This attribute would return scripts that can be run outside of the nix sandbox:

```
$(nix-build -A rocm-opencl-icd.impureTests.rocm-opencl)
```

Flakebi committed:
amdvlk: add nixos test

Flakebi commented:
Thanks a lot for your comments!

Good point about hydra, I was not sure if it will run all tests by default.
As you suggested, I added a new attribute `impureTests`, that runs the script, so `$(nix-build -A rocm-opencl-icd.impureTests)` works.

It’s still using a derivation to run the actual test, so that it’s _mostly_ isolated.
Most importantly, the test does not inherit the users environment variables and it runs in an isolated filesystem without the systems `/run/opengl-driver`.

@Madouura, I changed the `/dev/dri/card*` check to accept any card, good point. Get well soon!

[Diff](https://github.com/NixOS/nixpkgs/compare/7c5e0e3f100e2992deb4d652895b70f0e6d895fd..9c23648cb6938a5fdce13265ed36e0fc8ccc2a60) compared to the previous version.

Mic92 commented on nixos/tests/all-tests.nix:
```diff
@@ -75,6 +75,7 @@ in {
   airsonic = handleTest ./airsonic.nix {};
   allTerminfo = handleTest ./all-terminfo.nix {};
   amazon-init-shell = handleTest ./amazon-init-shell.nix {};
+  amdvlk = handleTest ./amdvlk.nix {};
```
This is better but I would still not include these tests in `all-tests.nix` since we include this in release.nix and therefore built them with hydra. Also those tests are not actually nixos vm tests like the others which is a bit confusing.
I have two solutions for that:

Put all those impure tests in `./pkgs/test` or just put them into the package directory along side the `default.nix` (my preference actually).

Mic92 commented on nixos/tests/make-test-hardware.nix:
```diff
@@ -0,0 +1,78 @@
+/* Create tests that run in the nix sandbox but get access to the host hardware
```
I actually would rename this to `make-impure-test.nix` since there might be other use cases not related to hardware that might be covered by this. I could imagine to have a sandbox-relaxed flag in the future to allow network access.

Mic92 commented on nixos/tests/rocm-opencl.nix:
```diff
@@ -0,0 +1,17 @@
+import ./make-test-hardware.nix ({ lib, pkgs, ... } : {
```
If we decide to move tests to individual package affected I would rather have a function like `makeImpureTest` that can be imported from nixpkgs and used outside of nixpkgs easier.

Flakebi commented:
Putting the tests under the packages is a good idea to keep `nixos/tests` clean.

I added `makeImpureTests` to `pkgs`, like you suggested.

Mic92 commented on pkgs/build-support/make-impure-test.nix:
```diff
@@ -0,0 +1,96 @@
+
+    ${prepareRunCommands}
+
+    /run/wrappers/bin/sudo nix-build --option extra-sandbox-paths '${sandboxPathsList}' ${lib.escapeShellArgs nixFlags} -A ${testPath} "$@"
```
It would be cool if sudo was skipped if the current user was a trusted user already, but I don't know what a good check for this be.

Mic92 replied:
Something like that might work:

```
$ nix show-config | grep 'trusted-users ='
trusted-users = joerg root nix
```

Mic92 replied:
But sudo might be also acceptable for the first version.

Flakebi replied:
Checking for user and group membership seems quite complicated to me as we need to parse and check it in bash.

Madouura replied:
IIRC the user needs to be in the ``video`` and/or ``render`` groups too.

Flakebi replied:
I think the video/render group does not apply to the tests because they run inside the sandbox without groups anyway.

The group is needed outside the sandbox to access `/dev/dri/card0` (`/dev/dri/renderD128` is world writable). To make `/dev/dri/card0` accessible inside the sandbox, the amdvlk test chmods it to be world-writable as well (and reverts that change afterwards).

Madouura commented on pkgs/development/libraries/rocm-opencl-icd/test.nix:
```diff
@@ -0,0 +1,19 @@
+  '';
+
+  meta = with lib.maintainers; {
+    maintainers = [ Flakebi ];
```
In an upcoming PR I'll be changing pretty much all the ROCm-related packages to ``teams.rocm.members``. (+ old maintainers not already in the team)

Flakebi commented:
Rebased and set `teams.rocm.members` as maintainer.

Madouura reviewed (approved):
Ran the tests, LGTM.
